cloudwatch-snowflake/main.tf

```diff
@@ -0,0 +1,78 @@
+data "archive_file" "zip" {
+  type        = "zip"
+  source_dir  = "${path.module}/tracker"
+  output_path = "${path.module}/dist/tracker.zip"
+}
+
+resource "aws_lambda_function" "cloudwatch_metrics_tracker" {
+  function_name = "cloudwatch_metrics_tracker"
+  description   = "Collect AWS Cloudwatch Metrics"
+  role          = var.lambda_role_arn == null ? aws_iam_role.lambda[0].arn : var.lambda_role_arn
+
+  filename = data.archive_file.zip.output_path
+  # Used to trigger updates
+  source_code_hash = data.archive_file.zip.output_base64sha256
+  handler          = "index.handler"
+  # List of available runtimes: https://docs.aws.amazon.com/lambda/latest/dg/API_CreateFunction.html#SSS-CreateFunction-request-Runtime
+  runtime = "nodejs16.x"
+  timeout = 300
+
+  environment {
+    variables = {
+      RESOURCES_DATA = jsonencode(var.tracked_resources_data)
+      PERIOD         = "3600"
+    }
+  }
+}
+
+resource "aws_iam_role" "lambda" {
+  count = var.lambda_role_arn == null ? 1 : 0
+  name  = "fivetran_lambda_${var.fivetran_group_id}"
+
+  assume_role_policy = jsonencode({
+    Version = "2012-10-17"
+    Statement = [
+      {
+        Action = "sts:AssumeRole"
+        Effect = "Allow"
+        Principal = {
+          Service = "lambda.amazonaws.com"
+        }
+      },
+      {
+        Action = "sts:AssumeRole"
+        Effect = "Allow"
+        Principal = {
+          AWS = "arn:aws:iam::${var.fivetran_aws_account_id}:root"
+        }
+        Condition = {
+          StringEquals = {
+            "sts:ExternalId" : "${var.fivetran_group_id}"
+          }
+        }
+      },
+    ]
+  })
+
+  inline_policy {
+    name = "LambdaInvokePolicy"
+
+    policy = jsonencode({
+      Version = "2012-10-17"
+      Statement = [
+        {
+          Sid      = "InvokePermission"
+          Effect   = "Allow"
+          Action   = ["lambda:InvokeFunction"]
+          Resource = "*"
```

Comment:
- couldn't/shouldn't this be precise?

Reply:
@rubyconvict  
Good catch. 
This could be precise but I think it shouldn't be, because
- This role is created for the purpose of connecting fivetran-lambda
- If we make it too specific (i.e. limited the access to only 1 specific lambda function), we would have to create 1 role for each lambda function that we create in the future (i.e. we will then have a lot of lambda role in the AWS IAM)
- I have made it an optional creation, so if user has created the lambda role somewhere, they can pass the lambda arn and skip the creation.

What do you think?

Reply:
Ok, I'm no expert at all, but not using '*' seems to be a popular opinion in general, so I had to mention it.
BTW, do we practice use of https://github.com/tenable/terrascan or anything like that check security concerns for us (at least running it locally, if not on CI)? CC: @swiknaba

Reply:
@rubyconvict do you want to write a short proposal for it? https://go/propose
Unsure how much sense it makes to run such scans on these modules, since many modules are flexible enough to use them securely or less securely. So probably these scans should be part of the CI pipeline of project-specific terraform scripts? -> if so, we can put that into our base CI script here: https://github.com/dbl-works/monorepo-scaffold ?